### PR TITLE
Propagate ephemeral drive mappings.

### PR DESCRIPTION
**Pull request by xcllnt**
When the source/guest AMI has a mapping for ephemeral drives, make
sure to propagate them to the soloized AMI. This makes sure that
they are attached as the same devices under solo.

Do not add drive mappings for ephemeral drives for no good reason,
especially when the mapping is the same as what AWS does by default.

### CODE REVIEW

File: brkt_cli/__init__.py [hunk 1]
```diff
@@ -633,14 +632,15 @@ def run(aws_svc, enc_svc_cls, image_id, encryptor_ami,
...
-            t.ephemeral_name = 'ephemeral%d' % (i,)
-            i += 1
-            new_bdm[drive] = t
+        # Propagate any ephemeral drive mappings to the soloized image
+        guest_image = aws_svc.get_all_images([image_id])[0]
+        guest_bdm = guest_image.block_device_mapping
+        for key in guest_bdm.keys():
+            guest_vol = guest_bdm[key]
+            if guest_vol.ephemeral_name:
+                log.info('Propagating block device mapping for %s at %s' %
```
Comment (bburtin) [inline brkt_cli/__init__.py]: Is this something that end users should see by default?  If not, we should make it a debug message.

File: brkt_cli/__init__.py [hunk 2]
```diff
@@ -633,14 +632,15 @@ def run(aws_svc, enc_svc_cls, image_id, encryptor_ami,
...
-        i = 0
-        # Just attach 4 ephemeral drives
-        # XXX Should get ephemeral drives from guest AMI (e.g. Centos 6.6)
-        for drive in ['/dev/sdb', '/dev/sdc', '/dev/sdd', '/dev/sde']:
-            t = BlockDeviceType()
-            t.ephemeral_name = 'ephemeral%d' % (i,)
-            i += 1
-            new_bdm[drive] = t
+        # Propagate any ephemeral drive mappings to the soloized image
+        guest_image = aws_svc.get_all_images([image_id])[0]
```
Comment (bburtin) [inline brkt_cli/__init__.py]: I don't see an implementation of get_all_images() in BaseAWSService.  Is this something that you added?  That would explain why the unit tests are failing.

Comment (xcllnt) [inline brkt_cli/__init__.py]: > On Oct 20, 2015, at 10:50 AM, Boris Burtin <notifications@github.com> wrote:
> 
> In brkt_cli/__init__.py <https://github.com/brkt/brkt-cli/pull/35#discussion_r42529294>:
> 
> > @@ -633,14 +632,15 @@ def run(aws_svc, enc_svc_cls, image_id, encryptor_ami,
> >          new_bdm['/dev/sda3'] = dev_log
> >          new_bdm['/dev/sda5'] = dev_guest_root
> >
> > -        i = 0
> > -        # Just attach 4 ephemeral drives
> > -        # XXX Should get ephemeral drives from guest AMI (e.g. Centos 6.6)
> > -        for drive in ['/dev/sdb', '/dev/sdc', '/dev/sdd', '/dev/sde']:
> > -            t = BlockDeviceType()
> > -            t.ephemeral_name = 'ephemeral%d' % (i,)
> > -            i += 1
> > -            new_bdm[drive] = t
> > +        # Propagate any ephemeral drive mappings to the soloized image
> > +        guest_image = aws_svc.get_all_images([image_id])[0]
> I don't see an implementation of get_all_images() in BaseAWSService. Is this something that you added? That would explain why the unit tests are failing.
> 
It’s a standard boto method.

--
Marcel Moolenaar
marcel@xcllnt.net <mailto:marcel@xcllnt.net>

Comment (bburtin) [inline brkt_cli/__init__.py]: This is the joy of untyped languages.  aws_svc is a BaseAWSService object, which we define.  There are two implementations: AWSService (production) and DummyAWSService (unit testing).  I'm confused that you're able to call get_all_images() at all, considering it's not defined in AWSService.

Comment (xcllnt) [inline brkt_cli/__init__.py]: > On Oct 20, 2015, at 11:22 AM, Boris Burtin <notifications@github.com> wrote:
> 
> In brkt_cli/__init__.py <https://github.com/brkt/brkt-cli/pull/35#discussion_r42533720>:
> 
> > @@ -633,14 +632,15 @@ def run(aws_svc, enc_svc_cls, image_id, encryptor_ami,
> >          new_bdm['/dev/sda3'] = dev_log
> >          new_bdm['/dev/sda5'] = dev_guest_root
> >
> > -        i = 0
> > -        # Just attach 4 ephemeral drives
> > -        # XXX Should get ephemeral drives from guest AMI (e.g. Centos 6.6)
> > -        for drive in ['/dev/sdb', '/dev/sdc', '/dev/sdd', '/dev/sde']:
> > -            t = BlockDeviceType()
> > -            t.ephemeral_name = 'ephemeral%d' % (i,)
> > -            i += 1
> > -            new_bdm[drive] = t
> > +        # Propagate any ephemeral drive mappings to the soloized image
> > +        guest_image = aws_svc.get_all_images([image_id])[0]
> This is the joy of untyped languages. aws_svc is a BaseAWSService object, which we define. There are two implementations: AWSService (production) and DummyAWSService (unit testing). I'm confused that you're able to call get_all_images() at all, considering it's not defined in AWSService.
> 
> 

I don’t know anything about the unit test, so I don’t know what the dummy implementation is supposed
to do. It’s not hard to have it return an object that has a block device mapping for an ephemeral drive,
but I’m sure we want to be able to have it return different things.

BTW: this was tested with the internal version...

--
Marcel Moolenaar
marcel@xcllnt.net <mailto:marcel@xcllnt.net>

**Comment by bburtin** [comment]
This commit conflicted with recent refactoring.  I merged the code in a separate commit.  Closing this pull request.

**Pull request closed without merging**